File: java/amp-rabbit/src/main/java/amp/rabbit/topology/Broker.java
package amp.rabbit.topology;

/**
 * 
 * @author Richard Clayton (Berico Technologies)
 */
public class Broker {

	/**
	 * This is used primarily to coach the ConnectionFactory
	 * to the credentials that should be used for this broker.
	 */
	protected String clusterId;
	protected String hostname;
	protected int port = 5672;
	protected String virtualHost = "/";
	protected String connectionStrategy;
	
	public Broker(){}
	
	public Broker(String hostname, int port) {
		
		this(null, hostname, port);
	}
	
	public Broker(String clusterId, String hostname, int port) {
		
		this(clusterId, hostname, port, "");
	}
	
	public Broker(String hostname, int port, String connectionStrategy) {
		
		this(null, hostname, port, connectionStrategy);
	}
	
	public Broker(String clusterId, String hostname, int port, String connectionStrategy) {
		
		this(clusterId, hostname, port, "/", connectionStrategy);
	}
	
	public Broker(String clusterId, String hostname, int port, String virtualhost, String connectionStrategy) {
		
		this.clusterId = clusterId;
		this.hostname = hostname;
		this.port = port;
		this.virtualHost = virtualhost;
		this.connectionStrategy = connectionStrategy;
	}

	public String getClusterId() {
		return clusterId;
	}
	
	public void setClusterId(String clusterId) {
		this.clusterId = clusterId;
	}
	
	public String getHostname() {
		return hostname;
	}
	
	public void setHostname(String hostname) {
		this.hostname = hostname;
	}
	
	public int getPort() {
		return port;
	}
	
	public void setPort(int port) {
		this.port = port;
	}

	public String getVirtualHost() {
		return virtualHost;
	}

	public void setVirtualHost(String virtualHost) {
		this.virtualHost = virtualHost;
	}
	
	public String getConnectionStrategy() {
		return connectionStrategy;
	}
	
	public void setConnectionStrategy(String strategy) {
		this.connectionStrategy = strategy;
	}
	
	@Override
	public String toString() {
		return "Broker [clusterId=" + clusterId + ", hostname=" + hostname
				+ ", port=" + port + ", vhost=" + virtualHost + ", connectionStrategy=" + connectionStrategy + "]";
	}
	
	@Override
	public int hashCode() {
		final int prime = 31;
		int result = 1;
		result = prime * result
				+ ((clusterId == null) ? 0 : clusterId.hashCode());
		result = prime * result
				+ ((hostname == null) ? 0 : hostname.hashCode());
		result = prime * result 
				+ ((connectionStrategy == null) ? 0 : connectionStrategy.hashCode());
		result = prime * result + port;
		result = prime * result 
				+ ((virtualHost == null) ? 0 : virtualHost.hashCode());
		return result;
	}

	@Override
	public boolean equals(Object obj) {
		if (this == obj)
			return true;
		if (obj == null)
			return false;
		if (getClass() != obj.getClass())
			return false;
		Broker other = (Broker) obj;
		if (clusterId == null) {
			if (other.clusterId != null)
				return false;
		} else if (!clusterId.equals(other.clusterId))
			return false;
		if (hostname == null) {
			if (other.hostname != null)
				return false;
		} else if (!hostname.equals(other.hostname))
			return false;
		if (connectionStrategy == null) {
			if (other.connectionStrategy != null)
				return false;
		} else if (!connectionStrategy.equals(other.connectionStrategy))
			return false;
		if (port != other.port)
			return false;
		if (virtualHost == null) {
			if (other.virtualHost != null)
				return false;
		} else if (!virtualHost.equals(other.virtualHost))
			return false;
		return true;
	}

	public static BrokerBuilder builder(){
		
		return new BrokerBuilder();
	}
	
	/**
	 * Simplifies construction of a Broker object.
	 * 
	 * @author Richard Clayton (Berico Technologies)
	 */
	public static class BrokerBuilder {
		
		Broker broker = new Broker();
		
		public BrokerBuilder cluster(String clusterId){
			
			broker.setClusterId(clusterId);
			
			return this;
		}
		
		public BrokerBuilder host(String host){
					
					broker.setHostname(host);
					
					return this;
				}
		
		public BrokerBuilder port(int port){
			
			broker.setPort(port);
			
			return this;
		}
		
		public BrokerBuilder vhost(String vhost){
			
			broker.setVirtualHost(vhost);
			
			return this;
		}
		
		public BrokerBuilder connectionStrategy(String strategy){
			
			broker.setConnectionStrategy(strategy);
			
			return this;
		}
		
		public Broker build(){
			
			return this.broker;
		}
	}
}

File: dotnet/source/amp.rabbit/topology/Broker.cs
namespace amp.rabbit.topology
{
    public class Broker
    {
        public string ClusterId { get; private set; }
        public string Hostname { get; private set; }
        public int Port { get; private set; }
        public string VirtualHost { get; private set; }
        public string ConnectionStrategy { get; private set; }

        public Broker(string hostname, int port)
            : this(null, hostname, port)
        {
        }

        public Broker(string clusterId, string hostname, int port)
            : this(clusterId, hostname, port, "")
        {
        }

        public Broker(string hostname, int port, string connectionStrategy)
            : this(null, hostname, port, "/", connectionStrategy)
        {
        }

        public Broker(string clusterId, string hostname, int port, string connectionStrategy)
            : this(clusterId, hostname, port, "/", connectionStrategy)
        {
        }

        public Broker(string clusterId, string hostname, int port, string virtualHost, string connectionStrategy)
        {
            ClusterId = clusterId;
            Hostname = hostname;
            Port = port;
            VirtualHost = virtualHost;
            ConnectionStrategy = connectionStrategy;
        }

        protected bool Equals(Broker other)
        {
            return string.Equals(ClusterId, other.ClusterId) 
                && string.Equals(Hostname, other.Hostname) 
                && Port == other.Port
                && string.Equals(VirtualHost, other.VirtualHost)
                && ConnectionStrategy.Equals(other.ConnectionStrategy);
        }

        public override string ToString()
        {
            return string.Format("Broker [ClusterId={0}, Hostname={1}, Port={2}, VHost={3}, ConnectionStrategy={4}]", 
                ClusterId, Hostname, Port, VirtualHost, ConnectionStrategy);
        }

        public override bool Equals(object obj)
        {
            if (ReferenceEquals(null, obj)) return false;
            if (ReferenceEquals(this, obj)) return true;
            if (obj.GetType() != this.GetType()) return false;
            return Equals((Broker) obj);
        }

        public override int GetHashCode()
        {
            unchecked
            {
                var hashCode = (ClusterId != null ? ClusterId.GetHashCode() : 0);
                hashCode = (hashCode*397) ^ (Hostname != null ? Hostname.GetHashCode() : 0);
                hashCode = (hashCode*397) ^ Port;
                hashCode = (hashCode * 397) ^ (VirtualHost != null ? VirtualHost.GetHashCode() : 0);
                hashCode = (hashCode * 397) ^ (ConnectionStrategy != null ? ConnectionStrategy.GetHashCode() : 0);
                return hashCode;
            }
        }
    }
}